frontend/tests/frontend.spec.js

```diff
@@ -0,0 +1,160 @@
+  expect(await window.title()).toBe("DedupliFHIR");
+});
+
+test("UI elements present", async () => {
+  // Check for content
+  const content = await window.locator("div.content");
+  expect(content).not.toBeNull();
+  expect(content.locator("h1")).toHaveText("DedupliFHIR");
+  await window.waitForSelector("h2");
+  expect(content.locator("h2")).toHaveText("Upload Patient Records File");
+
+  // Check for file dropper
+  const fileDropper = await window.locator("input[type='file']");
+  expect(fileDropper).not.toBeNull();
+  expect(fileDropper).toHaveId("file-input-specific");
+  expect(fileDropper).toHaveAttribute("accept", ".json,.csv");
+
+  // Check for dropdown
+  const dropdown = await window.locator("select");
+  expect(dropdown).not.toBeNull();
+  expect(dropdown).toHaveId("options");
+  expect(dropdown).toHaveClass("usa-select");
+
+  // Check for submit button
+  const submitButton = await window.locator("button[type='button']");
+  expect(submitButton).not.toBeNull();
+  expect(submitButton).toHaveId("submit");
+});
+
+test("submit without uploading file", async () => {
+  const submitButton = await window.locator("button[type='button']");
+  await submitButton.click();
+
+  const alert = await window.locator("#file-input-alert");
+  expect(alert).not.toBeNull();
+  expect(alert.locator("p")).toHaveText(
+    "File not found. Please upload a file.",
+  );
+});
+
+test("upload file and submit", async () => {
+  // TODO: remove this when height bug is fixed
+  await window.setViewportSize({ width: 1000, height: 3000 });
+
+  const fileDropper = await window.locator("input[type='file']");
+
+  await fileDropper.setInputFiles("../cli/deduplifhirLib/tests/test_data.csv");
+  // OR, put a custom test file in frontend/tests/
+  // await fileDropper.setInputFiles("tests/test_data.csv");
+
+  const submitButton = await window.locator("button[type='button']");
+  await submitButton.click();
+
+  await window.waitForSelector("div.loader");
+  const loader = await window.locator("div.loader");
+  expect(loader).not.toBeNull();
+});
+
+test("finished results present", async () => {
+  await window.waitForSelector("div.content");
+  const content = await window.locator("div.content");
+  expect(content).not.toBeNull();
+  expect(content.locator("h1")).toHaveText("DedupliFHIR");
+
+  // Wait for results
+  await window.waitForSelector("h2", {
+    timeout: 1000 * 60 * 2, // two minutes
+  });
+  expect(content.locator("h2")).toHaveText("Results");
+  expect(content.locator("a[href='../index.html']")).not.toBeNull();
+  expect(content.locator("#download")).not.toBeNull();
+
+  expect(content.locator("#download-success-alert")).toBeHidden();
+  expect(content.locator("#download-fail-alert")).toBeHidden();
+});
+
+test("download file", async () => {
+  const downloadButton = await window.locator("#download");
+  expect(downloadButton).not.toBeNull();
+
+  // cannot test download per https://github.com/microsoft/playwright/issues/5013
```

Comment:
@patsier-cms Any thoughts on how to go about testing the file download?

Reply:
I think the mock strategy in the Stack Overflow comment mentioned below should work? https://stackoverflow.com/a/75966734 Haven't tried it, but that seems reasonable

Reply:
I think mocking the download would work, but I'm just wondering if that would defeat the purpose of this test — we would be side-stepping the download which wouldn't be testing the download anymore, and this test would always pass by nature of side-stepping the download and always returning a "successful download" value

Reply:
That's a fair point and something we'll need to balance here. Ideally we can have the mock return the path to the file that should be generated, that way the only thing we would be skipping is file dialog.

At the same time, if we only mock `dialog.showSaveDialog` then we're still testing whether or not the dialog is triggered by the click. Either way we get a chance to test what happens after the download, so I think it's reasonable to include

Reply:
Added!